File: ce791hpc/hw4/water_supply_vector/water_supply.f
! program to perform enumeration for a water supply problem
	Program water_supply
        use ifport
		!use mpi
      include "mpif.h"
!declare variables
	integer, parameter :: ntrials = 10000000, iseed=567891
	real(8) x1(ntrials), x2(ntrials), x3(ntrials), cost(ntrials)
	real(8) sl1, sl2, sl3, hc1, hc2, hc3, demand, uc1, uc2, uc3, hardness
        real(8) time1, time2, time3, time4
        real(8) min_cost, time_obj, time_gen, elapsed_time, mflops
        integer min_cost_index

      	integer n, myid, numprocs, ierr, ntimes
		integer i,j,k
		real(8) min_cost_global

	sl1=25; sl2=120; sl3=100; hc1=200; hc2=2300; hc3=700
	demand=150; uc1=500; uc2=1000; uc3=2000

      call MPI_INIT(ierr)
      call MPI_COMM_RANK(MPI_COMM_WORLD, myid, ierr)
      call MPI_COMM_SIZE(MPI_COMM_WORLD, numprocs, ierr)

	  ! start timer - want to run for entire computation
      time1=MPI_Wtime()

! generate trial values
	!call wtime(time1)

	  ! outer loop to perform multiple trials
	  !ntimes=1024  ! too big to fit in int when multiplied by 3*n*nprocs
	  ntimes=10
      do k = 1,ntimes
!==================================
        j = 0
        call seed(iseed)
	do i=1,ntrials
	!do i=myid+1,ntrials,numprocs  
                x3(i) = -1
		do while (x3(i) > sl3 .or. x3(i) < 0 .or. &
                   hardness > 1200)
                   x1(i)=rand()*sl1
                   x2(i)=rand()*sl2
		   x3(i)=demand-x1(i)-x2(i)
		   hardness=(hc1*x1(i)+hc2*x2(i)+hc3*x3(i))/demand
                   j=j+1
                enddo
        enddo

	!    collect all the partial sums
! look into this later, MPI_Gather maybe
		  !call MPI_REDUCE(x1(i)uu,pi,1,MPI_DOUBLE_PRECISION,MPI_SUM,0, &
							   !MPI_COMM_WORLD,ierr)
	!call wtime(time2)
	!time_gen=time2-time1
!calculate cost
	do i=1,ntrials
	    cost(i)=uc1*x1(i)+uc2*x2(i)+uc3*x3(i)
        enddo
	!call wtime(time3)
	!time_obj=time3-time2
	min_cost=1e10
	do i=myid+1,ntrials,numprocs
            if (cost(i) < min_cost) then
               min_cost=cost(i)
               min_cost_index=i
            endif
        enddo

! use MPI_MIN on min_cost
call MPI_REDUCE(min_cost,min_cost_global,1,MPI_DOUBLE_PRECISION,MPI_MIN,0,MPI_COMM_WORLD,ierr)

	  ! end outer num trials loop
      enddo

	  ! end timer
      time2=MPI_Wtime()

	  ! calculate avg time per trial: time2-time1 gives total time
	  elapsed_time=(time2-time1)/ntimes

	!call wtime(time4)
	!elapsed_time=time4-time1

!    node 0 prints the answer.
      if (myid.eq.0) then
        !mflops_obj=5*ntrials*1e-6/time_obj
        mflops = dble(10*j+ntrials*5)*1e-6/elapsed_time

	!print *, 'amount drawn from source1 =',x1(min_cost_index)
	!print *, 'amount drawn from source2 =',x2(min_cost_index)
	!print *, 'amount drawn from source3 =',x3(min_cost_index)
	print *,'Minimum cost = ',nint(min_cost_global),' dollars'
	!print *,'Time to generate alternatives  = ',time_gen,' secs'
	!print *,'Time to calculate objective  = ',time_obj,' secs'
	!print *,'Mflop rating for Objective calculation = ',mflops_obj,' mflops'
	print *,'Mflop rating for entire code = ',mflops,' mflops'
	print *,'Total time = ',elapsed_time,' secs'
    
	endif
     call MPI_FINALIZE(ierr)
	end

File: ce791hpc/hw4/water_supply_vector/water_supply.java
/************************************************************
 analyze water supply options for a town
 exercise taken from CE297 (Dr. John Baugh)
************************************************************/

import java.io.*;
import java.util.*;
import java.lang.Runtime;

/*********************************************************** 
 uc = cost (in $) per million gallons per day (mgd)
             for each source
 supply limit = maximum mgd for each source
 hc = hc of water from each source (lb/mgd)
 hc_limit = maximum total hc per mgd
 demand = total demand in mgd
 time1 = wall clock time at the beginning of execution
 time2 = wall clock time at the end of execution
**************************************************************/

public class water_supply {

    public static void main(String args[]) {
       int noptions = 100000;
/* construct array objects*/
       double ad1 [] = new double [noptions];
       double ad2 [] = new double [noptions];
       double ad3 [] = new double [noptions];
       double cost[] = new double[noptions];
       Random generator = new Random();
// declare variables
       double uc1, uc2, uc3;
       double sl1, sl2, sl3;
       double hc1, hc2, hc3;
       double min_cost, avg_hc=0, hc_limit, demand;
       double time1, time2, time3, time4;
       double time_obj, time_gen, time_tot, mflops;
       int i, j, min_cost_index=0;
/* initialize constant parameters */
       uc1=500; uc2 = 1000; uc3=2000;
       sl1=25; sl2=120; sl3=100;
       hc1=200; hc2=2300; hc3=700;
       hc_limit=1200; demand = 150;
       time1=0; time2=0; time3=0; time4=0;
// place a call to System.gc to eliminate inconsistent timing 
//    due to garbage collection
       System.gc();
       time1= (double) System.currentTimeMillis();
//       System.out.println("time1 ="+time1);
/* initialize design variables subject to constraints */
       for (i=0; i<noptions; i++) {
          ad3[i] = 0;
          while (ad3[i] > sl3 ||
                 ad3[i] <= 0 ||
                 avg_hc >= hc_limit) {
                ad1[i]=generator.nextDouble()*sl1;
                ad2[i]=generator.nextDouble()*sl2;
                ad3[i]=demand-ad1[i]
                                   - ad2[i];
                avg_hc = (ad1[i]*hc1
                              +ad2[i]*hc2
                              +ad3[i]*hc3)/demand;
            }
        }
        time2= (double) System.currentTimeMillis();
//       System.out.println("time2 ="+time2);
/* time to generate alternatives */
        time_gen=(time2-time1)*1e-3;
/* calculate cost */
        min_cost = 1e10;
        for (i=0; i<noptions; i++) {
         cost[i] = uc1*ad1[i] 
                     +uc2*ad2[i]
                     +uc3*ad3[i];
        } 
        time3= (double) System.currentTimeMillis();
//        System.out.println("time3 ="+time3);
        time_obj=(time3-time2)*1e-3;
        for (i=0; i<noptions; i++) {
           if (cost[i] <= min_cost) {
               min_cost = cost[i];
               min_cost_index = i;
            } 
        }
        mflops = 5*noptions*1e-6/time_obj;
        time4= (double) System.currentTimeMillis();
//        System.out.println("time4 ="+time4);
        time_tot=(time4-time1)*1e-3;
        System.out.println("Amount drawn from source1 = "
                  +ad1[min_cost_index]+" mgd");
        System.out.println("Amount drawn from source2 = "
                  +ad2[min_cost_index]+" mgd");
        System.out.println("Amount drawn from source3 = "
                  +ad3[min_cost_index]+" mgd");
        System.out.println("Minimum cost = " + min_cost +" dollars");
        System.out.println("Time to generate alternatives  = "
           +time_gen+" secs");
        System.out.println("Time to calculate objective = "
           +time_obj+ " secs");
        System.out.println("Mflop rating for objective calculation = " 
           +mflops+"  mflops");
        System.out.println("Total time = "+time_tot+" secs");
    }

}
